### PR TITLE
Refine data reader and move data_reader to async_data_reader.

### LINKED ISSUE #705: We can improve the speed of DataReader for DeepASR model
Current implementation has much space to improve:
1. Use process instead of thread for batch assembling task.
2. Use process instead of thread for ordered data feeding task.
3. In batch iterator, should avoid to sharing raw numpy array between processes.

### PR DESCRIPTION
**Pull request by pkuyym**
Resolves #705

### CODE REVIEW

File: fluid/DeepASR/data_utils/util.py [hunk 1]
```diff
@@ -35,10 +37,118 @@ def lodtensor_to_ndarray(lod_tensor):
...
+
+class EpochEndSignal(object):
+    pass
+
+
 class CriticalException(Exception):
     pass
 
 
+class SharedNDArray(object):
```
Comment (kuke) [inline fluid/DeepASR/data_utils/util.py:79]: Please add some comments for this class

File: fluid/DeepASR/data_utils/util.py [hunk 2]
```diff
@@ -35,10 +37,118 @@ def lodtensor_to_ndarray(lod_tensor):
...
+                self._is_verify)
+
+    def __setstate__(self, state):
+        self._name = state[0]
+        self._inited = state[3]
+        self.zeros_like(state[1], state[2])
+        self._is_verify = state[4]
+
+
+class SharedMemoryPoolManager(object):
```
Comment (kuke) [inline fluid/DeepASR/data_utils/util.py]: Also please add some comments for this class.

File: fluid/DeepASR/train.py [hunk 3]
```diff
@@ -175,19 +175,21 @@ def test(exe):
...
-            feature_t.set(features, place)
-            feature_t.set_lod([lod])
-            label_t.set(labels, place)
-            label_t.set_lod([lod])
+            feature_t.set(features.ndarray, place)
+            feature_t.set_lod([lod.ndarray])
+            label_t.set(labels.ndarray, place)
+            label_t.set_lod([lod.ndarray])
+
+            test_data_reader.recycle(features, labels, lod)
```
Comment (kuke) [inline fluid/DeepASR/train.py:192]: Can we hide the recycle, e.g. move it into the iterator?

**Review by kuke — COMMENTED** [review]
Almost LGTM

**Review by zhxfl — COMMENTED** [review]
AG

**Review by zhxfl — COMMENTED** [review]
Almost LGTM

File: fluid/DeepASR/data_utils/util.py (hunk 1 above)
Comment (pkuyym) [reply fluid/DeepASR/data_utils/util.py:79]: Done.

File: fluid/DeepASR/data_utils/util.py (hunk 2 above)
Comment (pkuyym) [reply fluid/DeepASR/data_utils/util.py]: Done.

File: fluid/DeepASR/train.py (hunk 3 above)
Comment (pkuyym) [reply fluid/DeepASR/train.py:192]: I have considered to hide this function, seems not trivial. We can validate the correctness first.

File: fluid/DeepASR/data_utils/util.py [hunk 4]
```diff
@@ -35,10 +37,139 @@ def lodtensor_to_ndarray(lod_tensor):
...
+        is_verify (bool): Whether to do validation for writing operation.
+    """
+
+    def __init__(self, name, is_verify=False):
+        self._name = name
+        self._shm = None
+        self._buf = None
+        self._array = np.zeros(1, dtype=np.float32)
+        self._inited = False
+        self._is_verify = is_verify
```
Comment (kuke) [inline fluid/DeepASR/data_utils/util.py]: ```is_verify``` --> ```to_verify```

File: fluid/DeepASR/data_utils/util.py [hunk 5]
```diff
@@ -35,10 +37,139 @@ def lodtensor_to_ndarray(lod_tensor):
...
+    pass
+
+
 class CriticalException(Exception):
     pass
 
 
+class SharedNDArray(object):
+    """SharedNDArray utilizes shared memory to avoid data serialization when
+    object of which shared between different processes. We can reconstruct the
```
Comment (kuke) [inline fluid/DeepASR/data_utils/util.py]: ```object of which``` ---> ```data object```
```between``` --> ```among```

File: fluid/DeepASR/data_utils/util.py [hunk 6]
```diff
@@ -35,10 +37,139 @@ def lodtensor_to_ndarray(lod_tensor):
...
+    ndarray when memory address provided.
+
+    Args:
+        name (str): Address name of shared memory.
+        is_verify (bool): Whether to do validation for writing operation.
```
Comment (kuke) [inline fluid/DeepASR/data_utils/util.py]: ```is_verify``` --> ```to_verify```
```Whether to do validation for writing operation``` --> ```Whether to validate the writing operation```

File: fluid/DeepASR/data_utils/util.py (hunk 5 above)
Comment (pkuyym) [reply fluid/DeepASR/data_utils/util.py]: Done.

File: fluid/DeepASR/data_utils/util.py (hunk 4 above)
Comment (pkuyym) [reply fluid/DeepASR/data_utils/util.py]: Done.

File: fluid/DeepASR/data_utils/util.py (hunk 6 above)
Comment (pkuyym) [reply fluid/DeepASR/data_utils/util.py]: Done.

**Review by kuke — APPROVED** [review]
LGTM

**Pull request merged**